include/deal.II/numerics/vector_tools_evaluate.h

```diff
@@ -0,0 +1,302 @@
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii_vector_tools_evaluation_h
+#define dealii_vector_tools_evaluation_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/mpi_remote_point_evaluation.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/fe/fe_point_evaluation.h>
+
+#include <map>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace VectorTools
+{
+  /**
+   * Namespace for the flags for evaluate_at_points().
+   */
+  namespace EvaluationFlags
+  {
+    /**
+     * Flags for evaluate_at_points().
+     */
+    enum EvaluationFlags
+    {
+      /**
+       * Compute average.
+       */
+      avg = 0,
+      /**
+       * Compute maximum.
+       *
+       * @note Only available for scalar values.
+       */
+      max = 1,
+      /**
+       * Compute minimum.
+       *
+       * @note Only available for scalar values.
+       */
+      min = 2,
+      /**
+       * Take any value.
+       */
+      insert = 3
+    };
+  } // namespace EvaluationFlags
+
+  /**
+   * Given a (distributed) solution vector @p vector, evaluate the values at
+   * the (arbitrary and even remote) points specified by @p evaluation_points.
+   */
+  template <int n_components, int dim, int spacedim, typename VectorType>
+  std::vector<typename FEPointEvaluation<n_components, dim>::value_type>
+  evaluate_at_points(
+    const Mapping<dim> &                                  mapping,
+    const DoFHandler<dim, spacedim> &                     dof_handler,
+    const VectorType &                                    vector,
+    const std::vector<Point<spacedim>> &                  evaluation_points,
+    Utilities::MPI::RemotePointEvaluation<dim, spacedim> &cache,
+    const EvaluationFlags::EvaluationFlags flags = EvaluationFlags::avg);
+
+  /**
+   * Given a (distributed) solution vector @p vector, evaluate the values at
+   * the points specified by the @p cache which might have been set up by the
+   * above function.
+   *
+   * @note Refinement/coarsening/repartitioning leads to the invalidation of the
+   *   cache so that the above function has to be called again.
+   */
+  template <int n_components, int dim, int spacedim, typename VectorType>
+  std::vector<typename FEPointEvaluation<n_components, dim>::value_type>
+  evaluate_at_points(
+    const DoFHandler<dim, spacedim> &                           dof_handler,
+    const VectorType &                                          vector,
+    const Utilities::MPI::RemotePointEvaluation<dim, spacedim> &cache,
```

Comment:
I have been thinking about this function again and I think we should have the `cache` variable as the first argument, to avoid confusion with the previous function and the implicit `mapping` argument other functions use in deal.II.